### PR TITLE
add grid view option when choosing a file from within the app

### PR DESCRIPTION
**Pull request by apicellaj**
Implemented a grid view option for choosing a file from the app as mentioned in #185

### CODE REVIEW

**Comment by AndyScherzinger** [comment]
Hi @apicellaj 

thank you very much for your PR and welcome to Nextcloud. I did a small test and found two (minor) issues (for code review please see code comments):

* grid/list state information gets lost during screen rotation: switch to grid view and rotate device, the menu still say "grid view" instead of "list view"
* The grid view doesn't use the right layout file and grid width, see screenshot 1, while it should use the one the "regular" file listing view would use, see screenshot 2

Screenshot 1
device-2016-10-04-230925

Screenshot 2
device-2016-10-04-231417

File: src/com/owncloud/android/ui/activity/UploadFilesActivity.java [hunk 1]
```diff
@@ -223,6 +226,20 @@ public void onClick(DialogInterface dialog, int which) {
...
                 break;
             }
+            case R.id.action_switch_view: {
+                if (isGridView()) {
+                    item.setTitle(getString(R.string.action_switch_grid_view));
+                    item.setIcon(ContextCompat.getDrawable(getApplicationContext(),
+                            R.drawable.ic_view_module));
+                    mFileListFragment.switchToListView();
+                } else {
+                    item.setTitle(getApplicationContext().getString(R.string.action_switch_list_view));
```
Comment (AndyScherzinger) [inline src/com/owncloud/android/ui/activity/UploadFilesActivity.java]: ```getApplicationContext()``` can be removed

**Comment by apicellaj** [comment]
Hi @AndyScherzinger 

Thanks for the feedback and the warm welcome. The two issues seem to be resolved but let me know if you would like anything changed.

**Comment by AndyScherzinger** [comment]
@apicellaj thanks for the code changes! :+1: 

@tobiasKaminsky please have a look, now the grid view for local files would look like this:
device-2016-10-05-103545

--> The local files grid view does distinguish between grid-item and grid-image (image/video) which it can't at the moment since we would have to dynamically detect the mime type. That is possible but we do not have a implementation for that. So I would say it is out of scope and rather open another issue for this matter and declare this one feature complete. What do you think?

**Comment by AndyScherzinger** [comment]
Slight update, with some customization of the code:
```
@Override
    public View getView(int position, View convertView, ViewGroup parent) {
        View view = convertView;
        File file = null;
        boolean isGridView = true;
        LayoutInflater inflator = (LayoutInflater) mContext
                .getSystemService(Context.LAYOUT_INFLATER_SERVICE);

        if (mFiles != null && mFiles.length > position && mFiles[position] != null) {
            file = mFiles[position];
        }

        // Find out which layout should be displayed
        ViewType viewType;
        if (parent instanceof GridView) {
            String mimeType = MimetypeIconUtil.getBestMimeTypeByFilename(file.getName());
            if (file != null && (MimetypeIconUtil.isImage(mimeType) || MimetypeIconUtil.isVideo(mimeType))) {
                viewType = ViewType.GRID_IMAGE;
            } else {
                viewType = ViewType.GRID_ITEM;
            }
        } else {
            viewType = ViewType.LIST_ITEM;
            isGridView = false;
        }

        // create view only if differs, otherwise reuse
        if (convertView == null || convertView.getTag() != viewType) {
            switch (viewType) {
                case GRID_IMAGE:
                    view = inflator.inflate(R.layout.grid_image, parent, false);
                    view.setTag(ViewType.GRID_IMAGE);
                    break;
                case GRID_ITEM:
                    view = inflator.inflate(R.layout.grid_item, parent, false);
                    view.setTag(ViewType.GRID_ITEM);
                    break;
                case LIST_ITEM:
                    view = inflator.inflate(R.layout.list_item, parent, false);
                    view.setTag(ViewType.LIST_ITEM);
                    break;
            }
        }

        if (file != null) {
            if(!ViewType.GRID_IMAGE.equals(viewType)) {
                TextView fileName = (TextView) view.findViewById(R.id.Filename);
                String name = file.getName();
                fileName.setText(name);
            }
            ....
```

leads to this:
device-2016-10-05-111101

@tobiasKaminsky I would apply these changes _after_ we merged this PR since I did some refactorings in the background for MimeType utils and ViewType etc. etc.

**Comment by AndyScherzinger** [comment]
@tobiasKaminsky what do you think? I'd say merge + another PR later on. :+1:

**Comment by tobiasKaminsky** [comment]
@apicellaj thank you very much for your contribution!

**Comment by AndyScherzinger** [comment]
Thank you very much for your contribution @apicellaj :tada: It will be included in the next stable-feature release :rocket:

**Comment by apicellaj** [comment]
Great! Thanks again for your feedback

**Pull request merged**